tests/core/pyspec/eth2spec/test/helpers/optimistic_sync.py

```diff
@@ -0,0 +1,189 @@
+            return PayloadStatusV1StatusAlias.INVALIDATED
+
+
+@dataclass
+class PayloadStatusV1:
+    status: PayloadStatusV1Status = PayloadStatusV1Status.VALID
+    latest_valid_hash: Optional[Bytes32] = None
+    validation_error: Optional[str] = None
+
+    @property
+    def formatted_output(self):
+        return {
+            'status': str(self.status),
+            'latest_valid_hash': encode_hex(self.latest_valid_hash) if self.latest_valid_hash is not None else 'null',
+            'validation_error': str(self.validation_error) if self.validation_error is not None else 'null'
+        }
+
+
+class MegaStore(object):
+    spec = None
+    fc_store = None
+    opt_store = None
+    block_payload_statuses: Dict[Bytes32, PayloadStatusV1] = dict()
+
+    def __init__(self, spec, fc_store, opt_store):
+        self.spec = spec
+        self.fc_store = fc_store
+        self.opt_store = opt_store
+
+
+def get_optimistic_store(spec, anchor_state, anchor_block):
+    assert anchor_block.state_root == anchor_state.hash_tree_root()
+
+    opt_store = spec.OptimisticStore(
+        optimistic_roots=set(),
+        head_block_root=anchor_block.hash_tree_root(),
+
+    )
+    anchor_block_root = anchor_block.hash_tree_root()
+    opt_store.blocks[anchor_block_root] = anchor_block.copy()
+    opt_store.block_states[anchor_block_root] = anchor_state.copy()
+
+    return opt_store
+
+
+def add_optimistic_block(spec, mega_store, signed_block, test_steps,
+                         payload_status=None, status=PayloadStatusV1Status.SYNCING):
+    block = signed_block.message
+    block_root = block.hash_tree_root()
+    el_block_hash = block.body.execution_payload.block_hash
+
+    if payload_status is None:
+        payload_status = PayloadStatusV1(status=status)
+        if payload_status.status == PayloadStatusV1Status.VALID:
+            payload_status.latest_valid_hash = el_block_hash
+
+    mega_store.block_payload_statuses[block_root] = payload_status
+    test_steps.append({
+        'payload_status': payload_status.formatted_output,
+    })
+
+    # Optimistic sync
+
+    valid = True
+
+    # Case: INVALID
+    if payload_status.status == PayloadStatusV1Status.INVALID:
+        # Update parent status to INVALID
+        assert payload_status.latest_valid_hash is not None
+        current_block = block
+        while el_block_hash != payload_status.latest_valid_hash and el_block_hash != spec.Bytes32():
+            current_block_root = current_block.hash_tree_root()
+            assert current_block_root in mega_store.block_payload_statuses
+            mega_store.block_payload_statuses[current_block_root].status = PayloadStatusV1Status.INVALID
+            # Get parent
+            current_block = mega_store.fc_store.blocks[current_block.parent_root]
+            el_block_hash = current_block.body.execution_payload.block_hash
+
+    if payload_status.status != PayloadStatusV1Status.VALID:
+        valid = False
```

Comment:
Should the block be invalid (`valid = False`) only when the status does belong to `INVALIDATED` subset? I believe current logic will produce `valid: False` output in case when the status is e.g. `SYNCING`.

Reply:
Good point! So we may have to set it manually since `SYNCING` block's validity that we output here will be determined later in optimistic nodes.

I updated it in https://github.com/ethereum/consensus-specs/pull/2982/commits/ac717b106af596fc5758b0004b98badefcf18986, now it is set by parameter `valid` from the test script.